### DISCUSSION TITLE
Inverter Huwei - Modbus Proxy

### DISCUSSION THREAD

**Discussion by Criaga** [wlcrs/huawei_solar#743] [Connecting to the inverter]
Good morning, I have a Huawei inverter integrated into HA. 
I need to have a second modbus connection. 
Does anyone use "modbus Proxy"?
 Thank you

**Comment by Roving-Ronin** [comment]
https://github.com/wlcrs/huawei_solar/discussions?discussions_q=modbus+proxy+

**Reply by Criaga** [reply]
ok, I'm stupid 😅
some information, since I'm not very expert.
In the configuration do I have to put the IP of the inverter and also the port I use to integrate it into HA?
 I currently use port 6607.
 Thanks

**Reply by Roving-Ronin** [reply]
based upon what you've said, the Modbus Proxy config would look like this:

![image](https://github.com/wlcrs/huawei_solar/assets/108674933/84400e1a-825b-492d-ad6e-12fb3509c608)

**Reply by Criaga** [reply]
ok, I did everything as you suggested. 
Does Fusion Solar always include IP inverter and related port in the integration?

**Comment by mcamposv** [comment]
I built a lightweight Modbus TCP proxy running on an ESP32 specifically for Huawei setups — it lets
HA, EVCC, or any other client share the same inverter/EMMA connection without conflicts. No Docker
or extra server needed, just flash the ESP32. Repo here if it helps: 
https://github.com/mcamposv/modbusproxy

**Reply by Roving-Ronin** [reply]
@mcamposv  looks interesting, wondering what is needed to add support for other inverters/BESS? Using a Sigenergy myself nowadays that supports 5 connections, but know there's a lot of people with Sungrow / DEYE etc that also only support 1 connection. Assuming this supports all OEM rebranded Huawei such as iStore (i.e. doesn't look for Huawei or LUNA etc in the name and refuse connection if its iStore rebanded Huawei inverter and firmware) ?

Edit: Just looking and appears that all the web interface etc is in Spanish(?), not used to Arduinno code, but does it allow like ESPHome and custom components being able to use files defining the different text blocks in other languages ?

**Reply by mcamposv** [reply]
Great questions! To address both:

  Other inverters (Sungrow, DEYE, etc.): The proxy works purely at the Modbus TCP transport layer — it doesn't parse register contents, just multiplexes the single allowed connection to multiple clients. In
  principle it should work with any device that limits Modbus TCP connections to 1. That said, my setup is 100% Huawei so I haven't been able to test it with anything else — I actually created this project
  precisely because I couldn't find anything similar already out there.

  The one known limitation for other inverters is the startup connectivity check: on boot, the proxy queries a specific Modbus register to read the EMMA/SUN2000 device name and validate the connection is live.
  That check is Huawei-specific and would need to be either reparametrized or removed entirely for other inverters. Everything after that handshake is generic Modbus TCP proxying.

  If anyone tries it with Sungrow, DEYE, or similar and hits issues, I'd love to know — both to help debug and to make the startup check configurable for non-Huawei setups.

  OEM rebrands (iStore, etc.): Yes, fully transparent — the proxy doesn't inspect device names or firmware strings. If the underlying hardware speaks Modbus TCP, it should just work regardless of the brand on
  the box.

**Reply by Roving-Ronin** [reply]
@mcamposv  is the web interface only in spanish ? Can we get a selector / option to show in English ?

**Reply by Roving-Ronin** [reply]
What info do you need to add other brands ?

Sungrow device/model register is #4999 (register 5000 minus 1),   input_type = input. data_type = uint16.
SolarEdge device/model register is #40020 (was 40115),   input_type = holding. data_type = string. count = 16
Goodwe device/model register is #35011,  data_type = custom. count = 5.
Fronius device/model register is #40020,   input_type = holding. data_type = string. count = 16.


.
